src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1400,32 +1403,53 @@ private static class Key {
             File file;
             final boolean utf8;
 
-            public Key(File file, BasicFileAttributes attrs, ZipCoder zc) {
+            public Key(File file, BasicFileAttributes attrs, ZipCoder zc)
+                    throws IOException {
                 this.attrs = attrs;
-                this.file = file;
+                Object fk = attrs.fileKey();
+                if (fk != null) {
+                    // use file value, hashCode will use fileKey()
+                    this.file = file;
+                } else {
+                    this.file = getCanonicalFile(file);
+                }
                 this.utf8 = zc.isUTF8();
             }
 
+            @SuppressWarnings("removal")
+            private static File getCanonicalFile(File file) throws IOException {
+                try {
+                    return System.getSecurityManager() != null ?
+                        AccessController.doPrivileged((PrivilegedExceptionAction<File>) file::getCanonicalFile) :
+                        file.getCanonicalFile();
+                } catch (PrivilegedActionException e) {
+                    throw new IOException(e);
+                }
+            }
+
             public int hashCode() {
                 long t = utf8 ? 0 : Long.MAX_VALUE;
-                t += attrs.lastModifiedTime().toMillis();
-                return ((int)(t ^ (t >>> 32))) + file.hashCode();
+                Object fk = attrs.fileKey();
+                if (fk == null) {
+                    t += attrs.lastModifiedTime().toMillis();
+                }
+                return Long.hashCode(t) +
+                        (fk != null ? fk.hashCode() : file.hashCode());
             }
 
             public boolean equals(Object obj) {
                 if (obj instanceof Key key) {
                     if (key.utf8 != utf8) {
                         return false;
                     }
-                    if (!attrs.lastModifiedTime().equals(key.attrs.lastModifiedTime())) {
-                        return false;
-                    }
                     Object fk = attrs.fileKey();
                     if (fk != null) {
                         return fk.equals(key.attrs.fileKey());
-                    } else {
-                        return file.equals(key.file);
                     }
+                    if (!attrs.lastModifiedTime().equals(key.attrs.lastModifiedTime())) {
+                        return false;
+                    }
```

Comment:
The question in the previous comment is why is the last modified time used here, I think you have to dig into why it is used by equals.